File: mobile/ios/AutoRL/ModelRunner.swift
//
//  ModelRunner.swift
//  AutoRL
//
//  Singleton class for managing PyTorch model inference on iOS
//  Features: Thread-safe model loading, efficient image processing, performance monitoring
//

import Foundation
import UIKit
import Accelerate

#if canImport(PyTorchMobile)
import PyTorchMobile
#endif

actor ModelRunner {
    static let shared = ModelRunner()
    
    struct InferenceResult {
        let output: String
        let latencyMs: Int64
        let confidence: Float?
    }
    
    private var model: Any? = nil
    private var isModelLoaded = false
    
    private let modelName = "model_mobile_quant.pt"
    private let inputSize: Int = 224
    
    private init() {}
    
    /// Load the PyTorch model from bundle
    /// Thread-safe operation (actor isolation)
    func loadModel() async -> Bool {
        if isModelLoaded && model != nil {
            print("📱 Model already loaded")
            return true
        }
        
        do {
            print("📱 Loading model: \(modelName)")
            
            guard let modelPath = Bundle.main.path(forResource: modelName, ofType: nil) ??
                    Bundle.main.path(forResource: modelName.replacingOccurrences(of: ".pt", with: ""), ofType: "pt") else {
                print("❌ Model file not found: \(modelName)")
                return false
            }
            
            #if canImport(PyTorchMobile)
            // Note: PyTorch Mobile iOS API may differ - adjust based on actual implementation
            // let torchModel = try TorchModule(fileAtPath: modelPath)
            // model = torchModel
            print("⚠️ PyTorch Mobile import available but implementation needs PyTorch Mobile iOS SDK")
            isModelLoaded = true
            return true
            #else
            // Fallback: Use Core ML or mock implementation
            print("⚠️ PyTorch Mobile not available, using mock model")
            isModelLoaded = true
            return true
            #endif
        } catch {
            print("❌ Failed to load model: \(error)")
            isModelLoaded = false
            model = nil
            return false
        }
    }
    
    /// Check if model is loaded
    func isModelLoaded() -> Bool {
        return isModelLoaded && model != nil
    }
    
    /// Run inference on a UIImage
    func runInference(image: UIImage? = nil) async -> InferenceResult? {
        guard isModelLoaded, model != nil else {
            print("❌ Model not loaded")
            return nil
        }
        
        do {
            let startTime = Date()
            
            // Use provided image or load test image
            let inputImage = image ?? loadTestImage()
            guard let inputImage = inputImage else {
                print("❌ Could not load input image")
                return nil
            }
            
            // Preprocess image
            let inputTensor = try preprocessImage(inputImage)
            
            // Run inference
            #if canImport(PyTorchMobile)
            // Note: Actual PyTorch Mobile iOS API implementation needed
            // if let torchModel = model as? TorchModule {
            //     let outputTensor = try torchModel.forward(inputTensor)
            //     let result = postprocessOutput(outputTensor)
            //     let latencyMs = Int64(Date().timeIntervalSince(startTime) * 1000)
            //     return InferenceResult(output: result, latencyMs: latencyMs, confidence: extractConfidence(outputTensor))
            // }
            #endif
            
            // Mock inference for testing (fallback)
            let latencyMs = Int64(Date().timeIntervalSince(startTime) * 1000)
            return InferenceResult(
                output: "Detected: Button (score: 0.85) [Mock]",
                latencyMs: latencyMs,
                confidence: 0.85
            )
        } catch {
            print("❌ Inference failed: \(error)")
            return nil
        }
    }
    
    /// Preprocess UIImage to tensor
    private func preprocessImage(_ image: UIImage) throws -> Any {
        // Resize image to model input size
        guard let resizedImage = image.resized(to: CGSize(width: inputSize, height: inputSize)),
              let cgImage = resizedImage.cgImage else {
            throw NSError(domain: "ModelRunner", code: -1, userInfo: [NSLocalizedDescriptionKey: "Failed to resize image"])
        }
        
        // Convert to RGB
        let width = cgImage.width
        let height = cgImage.height
        let colorSpace = CGColorSpaceCreateDeviceRGB()
        let bytesPerPixel = 4
        let bytesPerRow = bytesPerPixel * width
        let bitsPerComponent = 8
        
        var pixelData = [UInt8](repeating: 0, count: width * height * bytesPerPixel)
        
        guard let context = CGContext(
            data: &pixelData,
            width: width,
            height: height,
            bitsPerComponent: bitsPerComponent,
            bytesPerRow: bytesPerRow,
            space: colorSpace,
            bitmapInfo: CGImageAlphaInfo.noneSkipLast.rawValue
        ) else {
            throw NSError(domain: "ModelRunner", code: -2, userInfo: [NSLocalizedDescriptionKey: "Failed to create context"])
        }
        
        context.draw(cgImage, in: CGRect(x: 0, y: 0, width: width, height: height))
        
        // Normalize pixel values to [0, 1] and apply ImageNet normalization
        let mean: [Float] = [0.485, 0.456, 0.406]
        let std: [Float] = [0.229, 0.224, 0.225]
        
        var normalizedData = [Float](repeating: 0, count: width * height * 3)
        
        for i in 0..<(width * height) {
            let r = Float(pixelData[i * 4]) / 255.0
            let g = Float(pixelData[i * 4 + 1]) / 255.0
            let b = Float(pixelData[i * 4 + 2]) / 255.0
            
            normalizedData[i] = (r - mean[0]) / std[0]
            normalizedData[i + width * height] = (g - mean[1]) / std[1]
            normalizedData[i + width * height * 2] = (b - mean[2]) / std[2]
        }
        
        #if canImport(PyTorchMobile)
        // Note: Actual PyTorch Mobile tensor creation API needed
        // let tensor = try IValue.from(normalizedData, size: [1, 3, Int32(height), Int32(width)])
        // return tensor
        #endif
        // Return normalized data for mock
        return normalizedData
    }
    
    /// Postprocess model output to readable result
    private func postprocessOutput(_ output: Any) -> String {
        #if canImport(PyTorchMobile)
        // Note: Actual PyTorch Mobile output parsing needed
        // guard let tensor = output as? IValue,
        //       let tensorData = try? tensor.toTensor().getDataAsFloatArray() else {
        //     return "Failed to parse output"
        // }
        // ... process tensorData ...
        #endif
        
        // Mock output
        let classNames = [
            "Button", "Text", "Image", "Input", "List",
            "Card", "Icon", "Header", "Footer", "Other"
        ]
        return "Detected: \(classNames[0]) (score: 0.85) [Mock]"
    }
    
    /// Extract confidence score from output
    private func extractConfidence(_ output: Any) -> Float? {
        #if canImport(PyTorchMobile)
        // Note: Actual PyTorch Mobile confidence extraction needed
        #endif
        return 0.85
    }
    
    /// Load test image from bundle
    private func loadTestImage() -> UIImage? {
        if let image = UIImage(named: "test_screen") {
            return image
        }
        
        // Create a placeholder image
        let size = CGSize(width: inputSize, height: inputSize)
        UIGraphicsBeginImageContextWithOptions(size, false, 1.0)
        UIColor.systemGray.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))
        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return image
    }
    
    /// Cleanup resources
    func cleanup() async {
        #if canImport(PyTorchMobile)
        // PyTorch models are automatically deallocated
        #endif
        model = nil
        isModelLoaded = false
        print("🧹 ModelRunner cleaned up")
    }
}

// MARK: - UIImage Extension
extension UIImage {
    func resized(to size: CGSize) -> UIImage? {
        UIGraphicsBeginImageContextWithOptions(size, false, scale)
        defer { UIGraphicsEndImageContext() }
        draw(in: CGRect(origin: .zero, size: size))
        return UIGraphicsGetImageFromCurrentImageContext()
    }
}

File: mobile/android/app/src/main/java/com/autorl/ModelRunner.kt
package com.autorl

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import dagger.hilt.android.qualifiers.ApplicationContext
import org.pytorch.IValue
import org.pytorch.Module
import org.pytorch.Tensor
import org.pytorch.torchvision.TensorImageUtils
import timber.log.Timber
import java.io.File
import java.io.FileOutputStream
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Singleton class for managing PyTorch model inference.
 * 
 * Features:
 * - Thread-safe model loading
 * - Efficient bitmap-to-tensor conversion
 * - Performance monitoring
 * - Resource cleanup
 * - Error handling
 */
@Singleton
class ModelRunner @Inject constructor(
    @ApplicationContext private val context: Context
) {
    
    data class InferenceResult(
        val output: String,
        val latencyMs: Long,
        val confidence: Float? = null
    )
    
    private var model: Module? = null
    private val modelLock = Any()
    private var isModelLoaded = false
    
    companion object {
        private const val MODEL_NAME = "model_mobile_quant.pt"
        private const val INPUT_SIZE = 224
        private const val TAG = "ModelRunner"
    }
    
    /**
     * Load the PyTorch model from assets.
     * Thread-safe operation.
     */
    fun loadModel(): Boolean {
        synchronized(modelLock) {
            if (isModelLoaded && model != null) {
                Timber.d("Model already loaded")
                return true
            }
            
            return try {
                Timber.d("Loading model: $MODEL_NAME")
                
                val modelPath = assetFilePath(context, MODEL_NAME)
                if (modelPath == null || !File(modelPath).exists()) {
                    Timber.e("Model file not found: $MODEL_NAME")
                    return false
                }
                
                model = Module.load(modelPath)
                isModelLoaded = true
                
                Timber.d("Model loaded successfully from: $modelPath")
                true
            } catch (e: Exception) {
                Timber.e(e, "Failed to load model")
                isModelLoaded = false
                model = null
                false
            }
        }
    }
    
    /**
     * Check if model is loaded.
     */
    fun isModelLoaded(): Boolean {
        synchronized(modelLock) {
            return isModelLoaded && model != null
        }
    }
    
    /**
     * Run inference on a bitmap image.
     */
    fun runInference(bitmap: Bitmap? = null): InferenceResult? {
        synchronized(modelLock) {
            if (!isModelLoaded || model == null) {
                Timber.e("Model not loaded")
                return null
            }
            
            return try {
                val startTime = System.currentTimeMillis()
                
                // Use provided bitmap or load test image
                val inputBitmap = bitmap ?: loadTestImage()
                if (inputBitmap == null) {
                    Timber.e("Could not load input image")
                    return null
                }
                
                // Preprocess image
                val inputTensor = preprocessBitmap(inputBitmap)
                
                // Run inference
                val outputTensor = model!!.forward(IValue.from(inputTensor)).toTensor()
                
                // Postprocess output
                val result = postprocessOutput(outputTensor)
                
                val latencyMs = System.currentTimeMillis() - startTime
                
                Timber.d("Inference completed in ${latencyMs}ms")
                
                InferenceResult(
                    output = result,
                    latencyMs = latencyMs,
                    confidence = extractConfidence(outputTensor)
                )
            } catch (e: Exception) {
                Timber.e(e, "Inference failed")
                null
            }
        }
    }
    
    /**
     * Preprocess bitmap to tensor.
     */
    private fun preprocessBitmap(bitmap: Bitmap): Tensor {
        // Resize bitmap to model input size
        val resizedBitmap = Bitmap.createScaledBitmap(
            bitmap,
            INPUT_SIZE,
            INPUT_SIZE,
            true
        )
        
        // Convert to tensor using PyTorch's utility
        return TensorImageUtils.bitmapToFloat32Tensor(
            resizedBitmap,
            TensorImageUtils.TORCHVISION_NORM_MEAN_RGB,
            TensorImageUtils.TORCHVISION_NORM_STD_RGB
        )
    }
    
    /**
     * Postprocess model output to readable result.
     */
    private fun postprocessOutput(outputTensor: Tensor): String {
        val scores = outputTensor.dataAsFloatArray
        
        // Find the class with highest score
        var maxIndex = 0
        var maxScore = scores[0]
        
        for (i in scores.indices) {
            if (scores[i] > maxScore) {
                maxScore = scores[i]
                maxIndex = i
            }
        }
        
        // Map index to class name (simplified - adjust based on your model)
        val classNames = arrayOf(
            "Button", "Text", "Image", "Input", "List",
            "Card", "Icon", "Header", "Footer", "Other"
        )
        
        val className = if (maxIndex < classNames.size) {
            classNames[maxIndex]
        } else {
            "Class_$maxIndex"
        }
        
        return "Detected: $className (score: ${String.format("%.2f", maxScore)})"
    }
    
    /**
     * Extract confidence score from output tensor.
     */
    private fun extractConfidence(outputTensor: Tensor): Float? {
        return try {
            val scores = outputTensor.dataAsFloatArray
            scores.maxOrNull()
        } catch (e: Exception) {
            Timber.e(e, "Failed to extract confidence")
            null
        }
    }
    
    /**
     * Load test image from assets.
     */
    private fun loadTestImage(): Bitmap? {
        return try {
            val inputStream = context.assets.open("test_screen.png")
            BitmapFactory.decodeStream(inputStream)
        } catch (e: Exception) {
            Timber.w(e, "Test image not found, using placeholder")
            // Create a placeholder bitmap
            Bitmap.createBitmap(INPUT_SIZE, INPUT_SIZE, Bitmap.Config.ARGB_8888)
        }
    }
    
    /**
     * Cleanup resources.
     */
    fun cleanup() {
        synchronized(modelLock) {
            model?.close()
            model = null
            isModelLoaded = false
            Timber.d("ModelRunner cleaned up")
        }
    }
}

